vpr/src/place/place_constraints.cpp

```diff
@@ -82,37 +82,95 @@ PartitionRegion constrained_macro_locs(const t_pl_macro& pl_macro) {
 
                 t_pl_loc min_pl_loc(reg_rect.xmin(), reg_rect.ymin(), block_regions[i].get_sub_tile());
 
-                t_pl_loc modified_min_pl_loc = min_pl_loc + offset;
+                t_pl_loc modified_min_pl_loc = min_pl_loc - offset;
 
                 t_pl_loc max_pl_loc(reg_rect.xmax(), reg_rect.ymax(), block_regions[i].get_sub_tile());
 
-                t_pl_loc modified_max_pl_loc = max_pl_loc + offset;
+                t_pl_loc modified_max_pl_loc = max_pl_loc - offset;
 
                 modified_reg.set_region_rect(modified_min_pl_loc.x, modified_min_pl_loc.y, modified_max_pl_loc.x, modified_max_pl_loc.y);
                 //check that subtile is not an invalid value before changing, otherwise it just stays -1
-                if (block_regions[i].get_sub_tile() != -1) {
+                if (block_regions[i].get_sub_tile() != NO_SUBTILE) {
                     modified_reg.set_sub_tile(modified_min_pl_loc.sub_tile);
                 }
 
                 modified_pr.add_to_part_region(modified_reg);
             }
 
             if (num_constrained_members == 1) {
-                macro_pr = modified_pr;
+                macro_head_pr = modified_pr;
             } else {
-                macro_pr = intersection(macro_pr, modified_pr);
+                macro_head_pr = intersection(macro_head_pr, modified_pr);
```

Comment:
Should also intersect the modified_pr with the grid dimensions to make sure it never goes outside the grid, and you get the tightest possible constraints.

Reply:
And even if some member of a macro has no constraint, you should consider it to have a constraint of the size of the grid, as it still implies a constraint for other members (if I'm the end of a vertical carry chain of length 2, it means the head of the chain can never be placed in the top row, even if I have no floorplan constraint on the second block in the chain).
If no member of a macro at all has a constraint, then you probably still want to avoid setting any region constraint as it might slow down the placer in that case.

Reply:
Yes, I'll update the function to have an intersection with the grid dimensions

Reply:
If a member of a macro has no constraint (when other members do), then it is updated to have a PartitionRegion that is the head macro, shifted by its offset. Basically, it is treated the same as the other macro members. This is because wherever the other members are constrained to, the non-constrained member will essentially have to be placed there too. 

If the macro has no members with floorplan constraints, then none of the updating PartitionRegion functions are called (update_macro_head_pr, update_macro_member_pr), so the PartitionRegions of all the members just remain empty.

Reply:
OK, that sounds good.